File: __tests__/extended_asset.test.cpp
/**
 *  @file
 *  @copyright defined in eosio.cdt/LICENSE.txt
 */

#include <string>

#include <eosio/tester.hpp>
#include <eosio/asset.hpp>

using std::string;

using eosio::name;
using eosio::symbol;
using eosio::extended_symbol;
using eosio::asset;
using eosio::extended_asset;

static constexpr int64_t asset_mask{(1LL << 62) - 1};
static constexpr int64_t asset_min{-asset_mask}; // -4611686018427387903
static constexpr int64_t asset_max{ asset_mask}; //  4611686018427387903

// Definitions in `eosio.cdt/libraries/eosio/asset.hpp`
EOSIO_TEST_BEGIN(extended_asset_type_test)
   static constexpr symbol sym_no_prec{"SYMBOLL",0};
   static constexpr symbol sym_prec{"SYMBOLL",63};

   static constexpr extended_symbol ext_sym_no_prec{sym_no_prec, name{"eosioaccountj"}};
   static constexpr extended_symbol ext_sym_prec{sym_prec, name{"eosioaccountj"}};

   static const asset asset_no_prec{0LL, sym_no_prec};
   static const asset asset_prec{0LL, sym_prec};

   //// extended_asset()
   CHECK_EQUAL( extended_asset{}.quantity, asset{} )
   CHECK_EQUAL( extended_asset{}.contract, name{}  )

   //// extended_asset(int64_t, extended_symbol)
   CHECK_EQUAL( (extended_asset{{},ext_sym_no_prec}.quantity), (asset{0LL, sym_no_prec}) )
   CHECK_EQUAL( (extended_asset{{},ext_sym_no_prec}.contract), (name{"eosioaccountj"}) )

   //// extended_asset(asset, name)
   CHECK_EQUAL( (extended_asset{asset_no_prec, name{"eosioaccountj"}}.quantity), (asset{ 0LL, sym_no_prec}) )
   CHECK_EQUAL( (extended_asset{asset_no_prec, name{"eosioaccountj"}}.contract), (name{"eosioaccountj"}) )

   // ------------------------------------------
   // extended_symbol get_extended_symbol()const
   CHECK_EQUAL( (extended_asset{{},ext_sym_no_prec}.get_extended_symbol()), (ext_sym_no_prec) )
   CHECK_EQUAL( (extended_asset{{},ext_sym_prec}.get_extended_symbol()), (ext_sym_prec) )

   // -----------------
   // void print()const
   CHECK_PRINT( "0 A@1", [](){extended_asset{asset{int64_t{0}, symbol{"A", 0}}, name{"1"}}.print();} )
   CHECK_PRINT( "0 A@5", [](){extended_asset{asset{int64_t{0}, symbol{"A", 0}}, name{"5"}}.print();} )
   CHECK_PRINT( "0 Z@a", [](){extended_asset{asset{int64_t{0}, symbol{"Z", 0}}, name{"a"}}.print();} )
   CHECK_PRINT( "0 Z@z", [](){extended_asset{asset{int64_t{0}, symbol{"Z", 0}}, name{"z"}}.print();} )

   CHECK_PRINT( "1.1 A@1", [](){extended_asset{asset{int64_t{11}, symbol{"A", 1}}, name{"1"}}.print();} )
   CHECK_PRINT( "1.1 A@5", [](){extended_asset{asset{int64_t{11}, symbol{"A", 1}}, name{"5"}}.print();} )
   CHECK_PRINT( "1.1 Z@a", [](){extended_asset{asset{int64_t{11}, symbol{"Z", 1}}, name{"a"}}.print();} )
   CHECK_PRINT( "1.1 Z@z", [](){extended_asset{asset{int64_t{11}, symbol{"Z", 1}}, name{"z"}}.print();} )

   CHECK_PRINT( "0.000000000000000000000000000000000000000000000000000000000000011 A@1",
      [](){extended_asset{asset{int64_t{11}, symbol{"A", 63}}, name{"1"}}.print();} )
   CHECK_PRINT( "0.000000000000000000000000000000000000000000000000000000000000011 A@5",
      [](){extended_asset{asset{int64_t{11}, symbol{"A", 63}}, name{"5"}}.print();} )
   CHECK_PRINT( "0.000000000000000000000000000000000000000000000000000000000000011 Z@a",
      [](){extended_asset{asset{int64_t{11}, symbol{"Z", 63}}, name{"a"}}.print();} )
   CHECK_PRINT( "0.000000000000000000000000000000000000000000000000000000000000011 Z@z",
      [](){extended_asset{asset{int64_t{11}, symbol{"Z", 63}}, name{"z"}}.print();} )

   CHECK_PRINT( "0 AAAAAAA@111111111111j", [](){extended_asset{asset{int64_t{0}, symbol{"AAAAAAA", 0}}, name{"111111111111j"}}.print();} )
   CHECK_PRINT( "0 AAAAAAA@555555555555j", [](){extended_asset{asset{int64_t{0}, symbol{"AAAAAAA", 0}}, name{"555555555555j"}}.print();} )
   CHECK_PRINT( "0 ZZZZZZZ@aaaaaaaaaaaaj", [](){extended_asset{asset{int64_t{0}, symbol{"ZZZZZZZ", 0}}, name{"aaaaaaaaaaaaj"}}.print();} )
   CHECK_PRINT( "0 ZZZZZZZ@zzzzzzzzzzzzj", [](){extended_asset{asset{int64_t{0}, symbol{"ZZZZZZZ", 0}}, name{"zzzzzzzzzzzzj"}}.print();} )

   CHECK_PRINT( "11 AAAAAAA@111111111111j", [](){extended_asset{asset{int64_t{11}, symbol{"AAAAAAA", 0}}, name{"111111111111j"}}.print();} )
   CHECK_PRINT( "11 AAAAAAA@555555555555j", [](){extended_asset{asset{int64_t{11}, symbol{"AAAAAAA", 0}}, name{"555555555555j"}}.print();} )
   CHECK_PRINT( "11 ZZZZZZZ@aaaaaaaaaaaaj", [](){extended_asset{asset{int64_t{11}, symbol{"ZZZZZZZ", 0}}, name{"aaaaaaaaaaaaj"}}.print();} )
   CHECK_PRINT( "11 ZZZZZZZ@zzzzzzzzzzzzj", [](){extended_asset{asset{int64_t{11}, symbol{"ZZZZZZZ", 0}}, name{"zzzzzzzzzzzzj"}}.print();} )

   CHECK_PRINT( "0.000000000000000000000000000000000000000000000000000000000000011 AAAAAAA@111111111111j",
     [](){extended_asset{asset{int64_t{11}, symbol{"AAAAAAA", 63}}, name{"111111111111j"}}.print();} )
   CHECK_PRINT( "0.000000000000000000000000000000000000000000000000000000000000011 AAAAAAA@555555555555j",
     [](){extended_asset{asset{int64_t{11}, symbol{"AAAAAAA", 63}}, name{"555555555555j"}}.print();} )
   CHECK_PRINT( "0.000000000000000000000000000000000000000000000000000000000000011 ZZZZZZZ@aaaaaaaaaaaaj",
     [](){extended_asset{asset{int64_t{11}, symbol{"ZZZZZZZ", 63}}, name{"aaaaaaaaaaaaj"}}.print();} )
   CHECK_PRINT( "0.000000000000000000000000000000000000000000000000000000000000011 ZZZZZZZ@zzzzzzzzzzzzj",
     [](){extended_asset{asset{int64_t{11}, symbol{"ZZZZZZZ", 63}}, name{"zzzzzzzzzzzzj"}}.print();} )

   // -------------------------------
   // extended_asset operator-()const
   CHECK_EQUAL( (-extended_asset{asset{ 0, sym_no_prec}, {}}.quantity), (extended_asset{asset_no_prec, {}}.quantity) )
   CHECK_EQUAL( (-extended_asset{asset{-0, sym_no_prec}, {}}.quantity), (extended_asset{asset_no_prec, {}}.quantity) )
   CHECK_EQUAL( (-extended_asset{asset{ 0, sym_prec}, {}}.quantity), (extended_asset{asset_prec, {}}.quantity) )
   CHECK_EQUAL( (-extended_asset{asset{-0, sym_prec}, {}}.quantity), (extended_asset{asset_prec, {}}.quantity) )

   CHECK_EQUAL( (-extended_asset{asset{1LL, sym_no_prec}, {}}.quantity), (extended_asset{asset{-1LL, sym_no_prec}, {}}.quantity) )
   CHECK_EQUAL( (-extended_asset{asset{1LL, sym_no_prec}, {}}.quantity), (extended_asset{asset{-1LL, sym_no_prec}, {}}.quantity) )
   CHECK_EQUAL( (-extended_asset{asset{1LL, sym_prec}, {}}.quantity), (extended_asset{asset{-1LL, sym_prec}, {}}.quantity) )
   CHECK_EQUAL( (-extended_asset{asset{1LL, sym_prec}, {}}.quantity), (extended_asset{asset{-1LL, sym_prec}, {}}.quantity) )

   CHECK_EQUAL( (-extended_asset{asset{asset_max, sym_no_prec}, {}}.quantity), (extended_asset{asset{asset_min, sym_no_prec}, {}}.quantity) )
   CHECK_EQUAL( (-extended_asset{asset{asset_max, sym_no_prec}, {}}.quantity), (extended_asset{asset{asset_min, sym_no_prec}, {}}.quantity) )
   CHECK_EQUAL( (-extended_asset{asset{asset_max, sym_prec}, {}}.quantity), (extended_asset{asset{asset_min, sym_prec}, {}}.quantity) )
   CHECK_EQUAL( (-extended_asset{asset{asset_max, sym_prec}, {}}.quantity), (extended_asset{asset{asset_min, sym_prec}, {}}.quantity) )

   // -----------------------------------------------------------------------------
   // friend extended_asset operator+(const extended_asset&, const extended_asset&)
   CHECK_EQUAL( (extended_asset{asset{0LL, sym_no_prec}, {}} + extended_asset{asset{ 0LL, sym_no_prec}, {}}), (extended_asset{asset_no_prec, {}}) )
   CHECK_EQUAL( (extended_asset{asset{1LL, sym_no_prec}, {}} + extended_asset{asset{-1LL, sym_no_prec}, {}}), (extended_asset{asset_no_prec, {}}) )

   CHECK_ASSERT( "type mismatch", (
      [&]() {
         extended_asset{asset_no_prec, name{"eosioaccountj"}} + extended_asset{asset_no_prec, name{"jtnuoccaoisoe"}};
      })
   )

   // -------------------------------------------------------------------------
   // friend extended_asset& operator+=(extended_asset&, const extended_asset&)
   extended_asset temp{asset_no_prec, {}};
   CHECK_EQUAL( (temp += temp), (extended_asset{asset_no_prec, {}}) )
   temp = extended_asset{asset{1LL, sym_no_prec}, {}};
   CHECK_EQUAL( (temp += extended_asset{asset{-1LL, sym_no_prec}, {}}), (extended_asset{asset_no_prec, {}}) )

   CHECK_ASSERT( "type mismatch", (
      [&]() {
         temp += extended_asset{asset_no_prec, name{"eosioaccountj"}};
      })
   )

   // -----------------------------------------------------------------------------
   // friend extended_asset operator-(const extended_asset&, const extended_asset&)
   CHECK_EQUAL( (extended_asset{asset_no_prec, {}} - extended_asset{asset_no_prec, {}}),
                  (extended_asset{asset_no_prec, {}}) )
   CHECK_EQUAL( (extended_asset{asset{1LL, sym_no_prec}, {}} - extended_asset{asset{1LL, sym_no_prec}, {}}),
                  (extended_asset{asset{asset_no_prec}, {}}) )

   CHECK_ASSERT( "type mismatch", (
      [&]() {
         extended_asset{asset_no_prec, name{"eosioaccountj"}} - extended_asset{asset_no_prec, name{"jtnuoccaoisoe"}};
      })
   )

   // --------------------------------------------------------------------------
   // friend extended_asset& operator-=(extended_asset&, const extended_asset&)
   temp = extended_asset{asset_no_prec, {}};
   CHECK_EQUAL( (temp -= temp), (extended_asset{asset_no_prec, {}}) )
   temp = extended_asset{asset{1LL, sym_no_prec}, {}};
   CHECK_EQUAL( (temp -= temp), (extended_asset{asset_no_prec, {}}) )

   CHECK_ASSERT( "type mismatch", (
      [&]() {
         temp -= extended_asset{asset_no_prec, name{"jtnuoccaoisoe"}};
      })
   )

   // --------------------------------------------------------------------
   // friend bool operator==(const extended_asset&, const extended_asset&)
   CHECK_EQUAL( (extended_asset{asset_no_prec, {}} == extended_asset{asset_no_prec, {}}), true )
   CHECK_EQUAL( (extended_asset{asset{1LL, sym_no_prec}, {}} == extended_asset{asset{1LL, sym_no_prec}, {}}), true )

   // --------------------------------------------------------------------
   // friend bool operator!=(const extended_asset&, const extended_asset&)
   CHECK_EQUAL( (extended_asset{asset_no_prec, name{"eosioaccountj"}} != extended_asset{asset_no_prec, name{"jtnuoccaoisoe"}}), true )
   CHECK_EQUAL( (extended_asset{asset{1LL, sym_no_prec}, {}} != extended_asset{asset{-1LL, sym_no_prec}, {}}), true )
   CHECK_EQUAL( (extended_asset{asset{1LL, sym_no_prec}, {}} != extended_asset{asset{ 0LL, sym_no_prec}, name{"eosioaccountj"}}), true )

   // -------------------------------------------------------------------
   // friend bool operator<(const extended_asset&, const extended_asset&)
   CHECK_EQUAL( (extended_asset{asset_no_prec, name{}} < extended_asset{asset{ 1LL, sym_no_prec}, {}}), true )
   CHECK_ASSERT( "type mismatch", (
      [&]() {
         bool b{extended_asset{{}, name{}} < extended_asset{{}, name{"eosioaccountj"}}};
         return b;
      })
   )

   // --------------------------------------------------------------------
   // friend bool operator<=(const extended_asset&, const extended_asset&)
   CHECK_EQUAL( (extended_asset{asset_no_prec, name{}} <= extended_asset{asset{ 1LL, sym_no_prec}, {}}), true );
   CHECK_ASSERT( "type mismatch", (
      [&]() {
         bool b{extended_asset{{}, name{}} <= extended_asset{{}, name{"eosioaccountj"}}};
         return b;
      })
   )

   // --------------------------------------------------------------------
   // friend bool operator>=(const extended_asset&, const extended_asset&)
   CHECK_EQUAL( (extended_asset{asset{ 1LL, sym_no_prec}, {}} >= extended_asset{asset_no_prec, name{}}), true );
   CHECK_ASSERT( "type mismatch", (
      [&]() {
         bool b{extended_asset{{}, name{}} >= extended_asset{{}, name{"eosioaccountj"}}};
         return b;
      })
   )
EOSIO_TEST_END

File: __tests__/extended_asset.test.ts
import { name, extended_symbol, extended_asset, asset, symbol, ExtendedAsset } from "..";
import bigInt from "big-integer";

const asset_mask = (bigInt(1).shiftLeft(62)).minus(1);
const asset_min = asset_mask.multiply(-1); // -4611686018427387903
const asset_max = asset_mask; //  4611686018427387903

test("asset::prevent mutation side effects", () => {
    const amount = bigInt(10000);
    const extQuantity = ExtendedAsset.from({contract: "eosio.token", quantity: "1.0000 EOS"});
    expect( extQuantity.quantity.amount ).toStrictEqual(amount);
    ExtendedAsset.times(extQuantity, 2);
    expect( extQuantity.quantity.amount ).toStrictEqual(amount);
    ExtendedAsset.div(extQuantity, 4);
    expect( extQuantity.quantity.amount ).toStrictEqual(amount);
    ExtendedAsset.plus(extQuantity, 500);
    expect( extQuantity.quantity.amount ).toStrictEqual(amount);
    ExtendedAsset.minus(extQuantity, 4000);
    expect( extQuantity.quantity.amount ).toStrictEqual(amount);
});

test("extended_symbol::error: type mismatch", () => {
    const error = "type mismatch";
    const a = ExtendedAsset.from({contract: "eosio.token", quantity: "1.0000 EOS"});
    const b = ExtendedAsset.from({contract: "foo.bar", quantity: "1.0000 EOS"});

    expect( () => ExtendedAsset.times(a, b) ).toThrow(error);
    expect( () => ExtendedAsset.times(b, a) ).toThrow(error);
    expect( () => a.times(b) ).toThrow(error);
    expect( () => b.times(a) ).toThrow(error);

    expect( () => ExtendedAsset.div(a, b) ).toThrow(error);
    expect( () => ExtendedAsset.div(b, a) ).toThrow(error);
    expect( () => a.div(b) ).toThrow(error);
    expect( () => b.div(a) ).toThrow(error);
});


test("extended_symbol::error: comparison of assets with different symbols is not allowed", () => {
    const error = "comparison of assets with different symbols is not allowed";
    const a = ExtendedAsset.from({contract: "eosio.token", quantity: "1.0000 EOS"});
    const c = ExtendedAsset.from({contract: "eosio.token", quantity: "1.0000 FOO"});

    expect( () => ExtendedAsset.times(a, c) ).toThrow(error);
    expect( () => ExtendedAsset.times(c, a) ).toThrow(error);
    expect( () => a.times(c) ).toThrow(error);
    expect( () => c.times(a) ).toThrow(error);

    expect( () => ExtendedAsset.div(a, c) ).toThrow(error);
    expect( () => ExtendedAsset.div(c, a) ).toThrow(error);
    expect( () => a.div(c) ).toThrow(error);
    expect( () => c.div(a) ).toThrow(error);
});

test("extended_symbol::from", () => {
    const ext_sym = ExtendedAsset.from({contract: "eosio.token", quantity: "1.0000 EOS"});
    expect( Number(ext_sym.quantity.amount) ).toBe(10000);
    expect( Number(ext_sym.quantity.symbol.precision()) ).toBe(4);
    expect( ext_sym.quantity.symbol.code().toString() ).toBe("EOS");
    expect( ext_sym.contract.toString() ).toBe("eosio.token");
});

test("extended_asset::json", () => {
    const ext_asset = extended_asset({contract: "eosio.token", quantity: "1.0000 EOS"});
    const { contract, quantity } = ext_asset.toJSON();

    expect( contract ).toStrictEqual("eosio.token");
    expect( quantity ).toStrictEqual("1.0000 EOS");
});

test("extended_asset", () => {
    const contract = name("eosio.token");
    const sym_no_prec = symbol("SYMBOLL", 0);
    const sym_prec = symbol("SYMBOLL", 63);

    const ext_sym_no_prec = extended_symbol( sym_no_prec, contract);
    const ext_sym_prec = extended_symbol( sym_prec, contract);

    const asset_no_prec = asset(0, sym_no_prec );
    const asset_prec = asset(0, sym_prec );

    //// extended_asset()
    expect( extended_asset().quantity ).toStrictEqual( asset() )
    expect( extended_asset().contract ).toStrictEqual( name() )

    //// extended_asset(int64_t, extended_symbol)
    expect( extended_asset( 0, ext_sym_no_prec ).quantity).toStrictEqual( asset(0, sym_no_prec))
    expect( extended_asset( 0, ext_sym_no_prec ).contract).toStrictEqual( contract)

    //// extended_asset(asset, name)
    expect( extended_asset( asset_no_prec, contract ).quantity).toStrictEqual( asset( 0, sym_no_prec ))
    expect( extended_asset( asset_no_prec, contract ).contract).toStrictEqual( contract)

    // ------------------------------------------
    // extended_symbol get_extended_symbol()const
    expect( extended_asset( ext_sym_no_prec ).get_extended_symbol()).toStrictEqual( ext_sym_no_prec )
    expect( extended_asset( ext_sym_prec ).get_extended_symbol()).toStrictEqual( ext_sym_prec )


    // -------------------------------
    // extended_asset operator-()const
    expect( extended_asset( asset(  0, sym_no_prec ) ).times( -1 ).quantity).toStrictEqual( extended_asset( asset_no_prec ).quantity)
    expect( extended_asset( asset( -0, sym_no_prec ) ).times( -1 ).quantity).toStrictEqual( extended_asset( asset_no_prec ).quantity)
    expect( extended_asset( asset(  0, sym_prec ) ).times( -1 ).quantity).toStrictEqual( extended_asset( asset_prec ).quantity)
    expect( extended_asset( asset( -0, sym_prec ) ).times( -1 ).quantity).toStrictEqual( extended_asset( asset_prec ).quantity)

    expect( extended_asset( asset( 1, sym_no_prec ) ).times( -1 ).quantity).toStrictEqual( extended_asset( asset( -1, sym_no_prec ) ).quantity)
    expect( extended_asset( asset( 1, sym_no_prec ) ).times( -1 ).quantity).toStrictEqual( extended_asset( asset( -1, sym_no_prec ) ).quantity)
    expect( extended_asset( asset( 1, sym_prec ) ).times( -1 ).quantity).toStrictEqual( extended_asset( asset( -1, sym_prec ) ).quantity)
    expect( extended_asset( asset( 1, sym_prec ) ).times( -1 ).quantity).toStrictEqual( extended_asset( asset( -1, sym_prec ) ).quantity)

    expect( extended_asset( asset( asset_max, sym_no_prec ) ).times( -1 ).quantity).toStrictEqual( extended_asset( asset( asset_min, sym_no_prec ) ).quantity)
    expect( extended_asset( asset( asset_max, sym_no_prec ) ).times( -1 ).quantity).toStrictEqual( extended_asset( asset( asset_min, sym_no_prec ) ).quantity)
    expect( extended_asset( asset( asset_max, sym_prec ) ).times( -1 ).quantity).toStrictEqual( extended_asset( asset( asset_min, sym_prec ) ).quantity)
    expect( extended_asset( asset( asset_max, sym_prec ) ).times( -1 ).quantity).toStrictEqual( extended_asset( asset( asset_min, sym_prec ) ).quantity)

    // -----------------------------------------------------------------------------
    // friend extended_asset operator+(const extended_asset&, const extended_asset&)
    expect( extended_asset( asset( 0, sym_no_prec ) ).plus( extended_asset( asset( 0, sym_no_prec ) ))).toStrictEqual( extended_asset( asset_no_prec ))
    expect( extended_asset( asset( 1, sym_no_prec ) ).plus( extended_asset( asset(-1, sym_no_prec ) ))).toStrictEqual( extended_asset( asset_no_prec ))

    expect( () => {
        extended_asset(asset_no_prec, contract ).plus( extended_asset( asset_no_prec, name("jtnuoccaoisoe") ))
    }).toThrow("type mismatch")

    // -------------------------------------------------------------------------
    // friend extended_asset& operator+=(extended_asset&, const extended_asset&)
    let temp = extended_asset( asset_no_prec );
    expect( temp.plus(temp) ).toStrictEqual( extended_asset( asset_no_prec ))
    temp = extended_asset( asset( 1, sym_no_prec) )
    expect( temp.plus( extended_asset( asset(-1, sym_no_prec )))).toStrictEqual(extended_asset( asset_no_prec ))

    expect( () => {
        temp.plus( extended_asset( asset_no_prec, contract ));
    }).toThrow("type mismatch")

    // -----------------------------------------------------------------------------
    // friend extended_asset operator-(const extended_asset&, const extended_asset&)
    expect( extended_asset( asset( 0, sym_no_prec ) ).minus( extended_asset( asset( 0, sym_no_prec ) ))).toStrictEqual( extended_asset( asset_no_prec ))
    expect( extended_asset( asset( 1, sym_no_prec ) ).minus( extended_asset( asset( 1, sym_no_prec ) ))).toStrictEqual( extended_asset( asset_no_prec ))

    expect( () => {
        extended_asset( asset_no_prec, contract ).minus( extended_asset( asset_no_prec, name("jtnuoccaoisoe") ))
    }).toThrow("type mismatch")

    // --------------------------------------------------------------------------
    // friend extended_asset& operator-=(extended_asset&, const extended_asset&)
    temp = extended_asset( asset_no_prec );
    expect( temp.minus(temp) ).toStrictEqual( extended_asset( asset_no_prec ))
    temp = extended_asset( asset( 1, sym_no_prec) )
    expect( temp.minus( extended_asset( asset( 1, sym_no_prec ) ))).toStrictEqual(extended_asset( asset_no_prec ))

    expect( () => {
        temp.minus( extended_asset( asset_no_prec, name("jtnuoccaoisoe") ));
    }).toThrow("type mismatch")

    // --------------------------------------------------------------------
    // friend bool operator==(const extended_asset&, const extended_asset&)
    expect( extended_asset( asset_no_prec ).isEqual( extended_asset( asset_no_prec ))).toBeTruthy();
    expect( extended_asset( asset(1, sym_no_prec) ).isEqual( extended_asset( asset(1, sym_no_prec ) ))).toBeTruthy()

    // --------------------------------------------------------------------
    // friend bool operator!=(const extended_asset&, const extended_asset&)
    expect( extended_asset( asset_no_prec, contract).isNotEqual( extended_asset( asset_no_prec, name("jtnuoccaoisoe")))).toBeTruthy()
    expect( extended_asset( asset(1, sym_no_prec) ).isNotEqual( extended_asset( asset( -1, sym_no_prec ) ))).toBeTruthy()
    expect( extended_asset( asset(1, sym_no_prec) ).isNotEqual( extended_asset( asset( 0, sym_no_prec ), contract ))).toBeTruthy();

    // -------------------------------------------------------------------
    // friend bool operator<(const extended_asset&, const extended_asset&)
    expect( extended_asset( asset_no_prec, name()).isLessThan( extended_asset( asset( 1, sym_no_prec ) ))).toBeTruthy();
    expect( () => {
        extended_asset( name() ).isLessThan( extended_asset( contract ));
    }).toThrow("type mismatch")

    // --------------------------------------------------------------------
    // friend bool operator<=(const extended_asset&, const extended_asset&)
    expect( extended_asset( asset_no_prec, name() ).isLessThanOrEqual( extended_asset( asset( 1, sym_no_prec)  ))).toBeTruthy();
    expect( () => {
        extended_asset( name() ).isLessThanOrEqual( extended_asset( contract ));
    }).toThrow("type mismatch")

    // --------------------------------------------------------------------
    // friend bool operator>=(const extended_asset&, const extended_asset&)
    expect( extended_asset( asset( 1, sym_no_prec ), name() ).isGreaterThanOrEqual( extended_asset( asset_no_prec, name()  ))).toBeTruthy();
    expect( () => {
        extended_asset( name() ).isGreaterThanOrEqual( extended_asset( contract ));
    }).toThrow("type mismatch")
});